connector/src/main/scala/com/microsoft/kusto/spark/utils/ContainerProvider.scala

```diff
@@ -163,60 +162,76 @@ class ContainerProvider(
 
 object ContainerProvider {
   private val className = this.getClass.getSimpleName
+  private val sasKeyCacheMap = new ConcurrentHashMap[String,ContainerAndSas]()
   def refreshUserSas(
       ingestionStorageParams: Array[IngestionStorageParameters],
+      isCacheExpired:Boolean,
       cacheExpirySeconds: Long,
       listPermissions:Boolean=false): ContainerAndSas = {
 
     val ingestionStorageParameter =
       IngestionStorageParameters.getRandomIngestionStorage(ingestionStorageParams)
 
-    if(StringUtils.isNotEmpty(ingestionStorageParameter.sas)){
-      KDSU.logInfo(className, "Using SAS token from ingestion storage parameter")
+    val key = ingestionStorageParameter.toString
+
+    // If the cache has not expired and the key is already in the cache, return the cached value
+    KDSU.logInfo("ContainerProvider",s" Checking cache for Key: $key")
+    if(!isCacheExpired && sasKeyCacheMap.contains(key)) {
       ContainerAndSas(
         s"${ingestionStorageParameter.storageUrl}/${ingestionStorageParameter.containerName}",
-        s"?${ingestionStorageParameter.sas}")
+        s"?${sasKeyCacheMap.get(key)}")
     } else {
-      KDSU.logInfo(className, s"Using user supplied ingestion storage $ingestionStorageParameter.Expires at " +
-        s"${OffsetDateTime.now.plusSeconds(cacheExpirySeconds)}")
-
-      val credential = if (StringUtils.isNotEmpty(ingestionStorageParameter.userMsi)) {
-        new ManagedIdentityCredentialBuilder().clientId(ingestionStorageParameter.userMsi).build()
+      if (StringUtils.isNotEmpty(ingestionStorageParameter.sas)) {
+        KDSU.logInfo(className, "Using SAS token from ingestion storage parameter")
+        ContainerAndSas(
+          s"${ingestionStorageParameter.storageUrl}/${ingestionStorageParameter.containerName}",
+          s"?${ingestionStorageParameter.sas}")
```

Comment:
as we are adding the "?"
Should we check on KDSU if fisrt character is "?" and if it is trim it ? or at least fail early?